woocommerce/sensei-certificates
# Add View Certificate Link to Course List templates

Part of https://github.com/Automattic/sensei/issues/6292

Also see https://github.com/Automattic/sensei/pull/6296

### Changes proposed in this Pull Request

Add "View Certificate" button to the Course List block templates.

### Notes

This was trickier than it seemed at first. In order to add the View Certificate button through a filter, we need to:

- Add the "View Certificate" block to the (locked) block template and the "allowed blocks" list of the Course Actions block. Note that the "View Certificate" block is actually a "button" block variation inside of a "buttons" block variation.
- In addition to the above (which allows the View Certificate block within the Course Actions block), we also need to filter the Course List block patterns (which is a PHP string containing the rendered output for the block content of each pattern) to include the View Certificate block with all of its appropriate styling.

This is a bit ugly tbh, but I'm not sure there a better way to do it. There are some implications of doing it like this:

- Since it's added to the pattern, if the Course List is already being used, the pattern will have to be reapplied in order to get the View Certificates blocks.
- Because the blocks are added unconditionally, and shown conditionally, there is a quirk: if the pattern is applied, and then the Certificates plugin is disabled, then the pattern shows a View Certificate button (which does nothing) on every Course in the list.
- That said, the assumption is that the Course List is probably not used in a lot of places on the site. The above problems should not be hard for users to fix if needed.

### Testing instructions

* Enable [this branch](https://github.com/Automattic/sensei/pull/6296) of Sensei (if the PR isn't merged yet) and this PR's branch of Certificates.
* Create some courses. Ensure that one is unstarted, one is started but not completed, one is completed without a certificate, and one is completed with a certificate.
* Create a page and add the Course List block.
* Try the different patterns for the block. For each one, ensure that the View Certificate button appears on the frontend only for the course which is completed and has a certificate.
* After testing, deactivate Certificates, remove the block, and add it again.
* As above, test with the different patterns, and ensure they work properly.

### Design notes

cc @Luchadores 

- For technical reasons, I adjusted the "List" pattern such that each button is the same width. This greatly simplified getting it to work well when adding the second button to come entries. Here is a screenshot, LMK what you think.

<img width="1173" alt="Screenshot 2022-12-14 at 16 44 35" src="https://user-images.githubusercontent.com/842193/207709959-8ec57259-a80c-4fb2-a381-170b63cc24b8.png">

---

- This is the "Grid" pattern. I don't really like how it looks with the two buttons. Any suggestions?

<img width="1380" alt="Screenshot 2022-12-14 at 16 44 06" src="https://user-images.githubusercontent.com/842193/207709871-fac0062d-0c82-464b-a600-161a6d1de151.png">
